improver/precipitation/precipitation_duration.py

```diff
@@ -0,0 +1,585 @@
+        and time as the three leading dimensions.
+
+        Args:
+            cube:
+                The cube to be structured.
+        Returns:
+            Tuple containing:
+                The reordered cube and the threshold coordinate.
+        """
+        # Pick out the threshold coordinates in the inputs.
+        thresh = cube.coord(var_name="threshold")
+        # Promote any scalar threshold coordinates to dimensions so that the
+        # array shapes are as expected.
+        if thresh not in cube.coords(dim_coords=True):
+            cube = new_axis(cube, thresh)
+        # Likewise, ensure the time coordinate is a dimension coordinate in
+        # case a period equal to the target period is passed in.
+        time_crd = cube.coord("time")
+        if time_crd not in cube.coords(dim_coords=True):
+            cube = new_axis(cube, time_crd)
+        # Enforce a coordinate order so we know which dimensions are which
+        # when indexing the data arrays.
+        enforce_coordinate_ordering(cube, ["realization", thresh.name(), "time"])
+        return cube, thresh
+
+    def _create_output_cube(
+        self,
+        data: ndarray,
+        cubes: Tuple[Cube, Cube],
+        acc_thresh: DimCoord,
+        rate_thresh: DimCoord,
+        spatial_dim_indices,
+    ) -> Cube:
+        """Create an output cube for the final diagnostic, which is the
+        set of percentiles describing the fraction of the constructed
+        period that has been classified as wet using the various accumulation
+        and peak rate thresholds.
+
+        Args:
+            data:
+                The percentile data to be stored in the cube.
+            cubes:
+                Cubes from which to generate attributes, and the first of
+                which is used as a template from which to take time and
+                spatial coordinates.
+            acc_thresh:
+                The accumulation threshold to add to the cube.
+            rate_thresh:
+                The rate threshold to add to the cube.
+            spatial_dim_indices:
+                Indices of spatial coordinates on the the input cubes.
+                This allows us to select the correct dimensions when handling
+                either gridded or spot inputs.
+        Returns:
+            A Cube with a suitable name and metadata for the diagnostic data
+            stored within it.
+        """
+        percentile_coord = DimCoord(self.percentiles, long_name="percentile", units="%")
+
+        mandatory_attributes = generate_mandatory_attributes(
+            cubes, model_id_attr=self.model_id_attr
+        )
+        mandatory_attributes["precipitation_sampling_period_in_hours"] = self.period
+
+        time_coords = [cubes[0].coord("forecast_reference_time").copy()]
+        for crd in ["time", "forecast_period"]:
+            time_crd = cubes[0].coord(crd).copy()
+            time_coords.append(time_crd.collapsed())
```

Comment:
Just to be sure, this results in a `time` coordinate and `forecast_period` coordinate with a point that is in the middle of the period. This is different to the typical IMPROVER convention. You've also added a `collapse_time` function in this PR, but I don't think that it's used. Should it be used here?

Reply:
Definitely not my intention. Good spot. I've moved this statement down to where it's applied, added it to the cube, and then used the enforcement utility to shift the points to the end. The `collapse_time` utility is no longer used in this code, but I'm using it elsewhere, so it can still go in with this if you don't object.